modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridQueryOptimizer.java

```diff
@@ -0,0 +1,865 @@
+
+    /** Predicat returns {@code true} if current element is an INNER JOIN. */
+    private static final Predicate<GridSqlAst> ELEMENT_IS_INNER_JOIN
+        = elem -> elem instanceof GridSqlJoin && !((GridSqlJoin)elem).isLeftOuter();
+
+    /** Predicat returns {@code true} if current element is an EQUAL or EQUAL_NULL_SAFE operation. */
+    private static final Predicate<GridSqlAst> ELEMENT_IS_EQ
+        = elem -> elem instanceof GridSqlOperation && (EQUAL == ((GridSqlOperation)elem).operationType()
+        || EQUAL_NULL_SAFE == ((GridSqlOperation)elem).operationType());
+
+    /**
+     * Whether to apply optimization or not.
+     */
+    private static volatile Boolean optimizationEnabled;
+
+    /**
+     * @return {@code true} if optimization should be applied.
+     */
+    @SuppressWarnings("NonThreadSafeLazyInitialization")
+    private static boolean optimizationEnabled() {
+        if (optimizationEnabled == null) { // it's OK if this will be initialized several times in case of races
+            optimizationEnabled = Boolean.parseBoolean(
+                System.getProperty(IgniteSystemProperties.IGNITE_ENABLE_SUBQUERY_REWRITE_OPTIMIZATION, "true")
+            );
+        }
+
+        return optimizationEnabled;
+    }
+
+    /**
+     * @param parent Parent.
+     */
+    public static boolean pullOutSubQueries(GridSqlQuery parent) {
+        boolean wasPulledOut = false;
+
+        if (!optimizationEnabled())
+            return wasPulledOut;
+
+        if (parent instanceof GridSqlUnion) {
+            GridSqlUnion union = (GridSqlUnion)parent;
+
+            wasPulledOut |= pullOutSubQueries(union.left());
+            wasPulledOut |= pullOutSubQueries(union.right());
+
+            return wasPulledOut;
+        }
+
+        // pull out select expressions from SELECT clause
+        GridSqlSelect select = (GridSqlSelect)parent;
+        boolean locPulledOut;
+
+        for (int i = 0; i < select.allColumns(); i++) {
+            locPulledOut = false;
+            GridSqlAst col = select.columns(false).get(i);
+
+            if (col instanceof GridSqlSubquery)
+                locPulledOut = pullOutSubQryFromSelectExpr(select, null, i);
+            else {
+                ASTNodeFinder finder = new ASTNodeFinder(
+                    col,
+                    ELEMENT_WITH_SUBQUERY
+                );
+
+                ASTNodeFinder.Result res;
+                while ((res = finder.findNext()) != null)
+                    locPulledOut |= pullOutSubQryFromSelectExpr(select, res.getEl(), res.getIdx());
+            }
+
+            if (locPulledOut) // we have to analize just pulled out element as well
+                i--;
+
+            wasPulledOut |= locPulledOut;
+        }
+
+        // pull out sub-queries from table list
+        do {
+            locPulledOut = false;
+
+            if (ELEMENT_WITH_ALIAS_WITH_SUBQUERY.test(null, select.from()))
```

Comment:
Does it make sense to check this condition at each iteration (and the same condition at the loops below)?
Please clarify. It's not clear to me.

Reply:
No, it doesn't. Good catch!

Reply:
well, after some investigation i'd decided to get rid of this limitation, so now it does make sense. I left a comment describing the reason in the code